Fix unpack_array_vector to produce an array of bytes

tautschnig opened the pull request:
We wrongly created a return type (though not the underlying expression,
which was correct!) as an array of elements of the original subtype,
even though the sole purpose of unpack_array_vector is to turn the
original subtype into a sequence of bytes.

We seemingly neither had a test covering this code nor did we have
additional result verification in place. Both are now rectified: a new
regression test is added, and an invariant is checked right after
returning from unpack_rec (which may invoke unpack_array_vector).

Fixes: #5718

<!---
Thank you for your contribution. Please make sure your pull request fulfils all of the below requirements. If you cannot currently tick all the boxes, but would still like to create a PR, then add the label "work in progress" and assign the PR to yourself.
--->

- [x] Each commit message has a non-empty body, explaining why the change was made.
- n/a Methods or procedures I have added are documented, following the guidelines provided in CODING_STANDARD.md.
- n/a The feature or user visible behaviour I have added or modified has been documented in the User Guide in doc/cprover-manual/
- [x] Regression or unit tests are included, or existing tests cover the modified code (in this case I have detailed which ones those are in the commit message).
- n/a My commit message includes data points confirming performance improvements (if claimed).
- [x] My PR is restricted to a single feature or bugfix.
- n/a White-space or formatting changes outside the feature-related changed lines are in commits of their own.

<!---
See, e.g., https://chris.beams.io/posts/git-commit/ for general guidelines on commit messages.

If you have created commits mixing multiple features and/or unrelated white-space changes, use a sequence involving git reset and git add -p to fix this.
--->

tautschnig committed:
Fix unpack_array_vector to produce an array of bytes

We wrongly created a return type (though not the underlying expression,
which was correct!) as an array of elements of the original subtype,
even though the sole purpose of unpack_array_vector is to turn the
original subtype into a sequence of bytes.

We seemingly neither had a test covering this code nor did we have
additional result verification in place. Both are now rectified: a new
regression test is added, and an invariant is checked right after
returning from unpack_rec (which may invoke unpack_array_vector).

Fixes: #5718

hannes-steffenhagen-diffblue commented on src/solvers/lowering/byte_operators.cpp:
```diff
@@ -475,26 +475,17 @@ static exprt unpack_array_vector(
-    }
+    const exprt array_vector_size = src.type().id() == ID_vector
+                                      ? to_vector_type(src.type()).size()
+                                      : to_array_type(src.type()).size();
```
Good refactor, but would prefer if this was a separate commit

tautschnig replied:
Done!

hannes-steffenhagen-diffblue commented on src/solvers/lowering/byte_operators.cpp:
```diff
@@ -475,26 +475,17 @@ static exprt unpack_array_vector(
-        *subtype,
-        mult_exprt{*array_vector_size,
-                   from_integer(el_bytes, array_vector_size->type())}}};
+        bv_typet{8},
```
Yep, can independently confirm this was the problem here

hannes-steffenhagen-diffblue commented on src/solvers/lowering/byte_operators.cpp:
```diff
@@ -1034,6 +1025,9 @@ exprt lower_byte_extract(const byte_extract_exprt &src, const namespacet &ns)
     src.op(), little_endian, lower_bound_int_opt, upper_bound_int_opt, ns);
+  CHECK_RETURN(
+    to_bitvector_type(to_type_with_subtype(unpacked.op().type()).subtype())
+      .get_width() == 8);
```
Good call, should stop a similar issue cropping up going forward a bit earlier.

That said, I'm a bit worried about all the hard-coded 8s and 7s in this module

tautschnig replied:
Ack, I really should revive #917, which would sort this out.

hannes-steffenhagen-diffblue commented on regression/cbmc-library/memcpy-08/main.c:
```diff
@@ -0,0 +1,22 @@
+  __CPROVER_assume(size >= 2);
+  // limit the size to ensure size * sizeof(int) wouldn't overflow, and we don't
+  // reach the maximum object size
+  __CPROVER_assume(size <= 1000);
```
Good find, I couldn't figure out what exactly was causing the array comprehension to be emitted here. I had considered just writing a unit test for this particular case instead.

tautschnig committed:
unpack_array_vector: Replace now-unnecessary if/else by ?:

We only need to set a single variable after the preceding fix, and can
thus get away without an optionalt.
